src/repositories/gitlab_updater.rs

```diff
@@ -0,0 +1,323 @@
+        &self,
+        host: &str,
+        query: &str,
+        variables: impl serde::Serialize,
+    ) -> Result<GraphResponse<T>> {
+        Ok(self
+            .client
+            .post(&format!("https://{}/api/graphql", host))
+            .json(&serde_json::json!({
+                "query": query,
+                "variables": variables,
+            }))
+            .send()?
+            .error_for_status()?
+            .json()?)
+    }
+}
+
+#[derive(Debug, failure::Fail)]
+#[fail(display = "rate limit reached")]
+struct RateLimitReached;
+
+#[derive(Debug, Deserialize)]
+struct GraphProjects<T> {
+    projects: GraphNodes<T>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphResponse<T> {
+    data: Option<T>,
+    #[serde(default)]
+    errors: Vec<GraphError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphError {
+    message: String,
+    locations: Vec<GraphErrorLocation>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphErrorLocation {
+    line: u32,
+    column: u32,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphRateLimit {
+    remaining: u32,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GraphNodes<T> {
+    nodes: Vec<T>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphProjectNode {
+    project: Option<GraphProject>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GraphProject {
+    id: String,
+    name_with_namespace: String,
+    last_activity_at: Option<DateTime<Utc>>,
+    description: Option<String>,
+    star_count: i64,
+    forks_count: i64,
+    open_issues_count: Option<i64>,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::repositories::RepositoryName;
+
+    #[test]
+    fn test_repository_name() {
+        macro_rules! assert_name {
+            ($url:expr => ($owner:expr, $repo:expr, $host:expr)) => {
+                assert_eq!(
+                    GitlabUpdater::repository_name($url),
+                    Some(RepositoryName {
+                        owner: $owner,
+                        repo: $repo,
+                        host: $host,
+                    })
+                );
+            };
+            ($url:expr => None) => {
+                assert_eq!(GitlabUpdater::repository_name($url), None);
+            };
+        }
+
+        assert_name!("https://gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi", "gitlab.com"));
+        assert_name!("http://gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi", "gitlab.com"));
+        assert_name!("https://gitlab.com/onur/cratesfyi.git" => ("onur", "cratesfyi", "gitlab.com"));
+        assert_name!("https://gitlab.com/docopt/docopt.rs" => ("docopt", "docopt.rs", "gitlab.com"));
+        assert_name!("https://gitlab.com/onur23cmD_M_R_L_/crates_fy-i" => (
+            "onur23cmD_M_R_L_", "crates_fy-i", "gitlab.com"
+        ));
+        assert_name!("https://gitlab.freedesktop.org/test/test" => (
+            "test", "test", "gitlab.freedesktop.org"
+        ));
```

Comment:
I just remembered that gitlab allows nesting groups, https://gitlab.com/nemo157-testing/subgroup/testing, there should be a testcase for this too.

Reply:
I'll need more information to be able to handle that: in here, the repository is `testing` and the owner is `nemo157-testing`?

Reply:
I guess the owner is `nemo157-testing/subgroup`? I've not had to hit the gitlab API with any nested groups yet so not sure exactly how it handles it.

Reply:
Yeah, it needs all the segments:

```console
> curl -s https://gitlab.com/api/graphql -H 'Content-Type: application/json' --data '{
  "query":"query($fullPath: ID!){ project(fullPath: $fullPath){id,nameWithNamespace}}",
  "variables":{"fullPath":"nemo157-testing/subgroup/testing"}
}' | jq

{
  "data": {
    "project": {
      "id": "gid://gitlab/Project/24037584",
      "nameWithNamespace": "nemo157-testing / subgroup / testing"
    }
  }
}
```